### ISSUE TITLE
Question: How does top-k expert routing work during inference?

### ISSUE THREAD

**Issue by maxyu1115** [raymin0223/mixture_of_recursions#11]
Hi there, as title, I've been trying to understand how the top-k expert routing works during inference. Sorry in advance if I just missed this info when reading through the paper.

From what I understand, for an initial prompt of length $n$ and recursion depth 3, we would first prefill with $n$ tokens at depth 1, top $\frac{2n}{3}$ at depth 2, and top $\frac{n}{3}$ at depth 3. We would then auto-regressively generate token by token.

This is where I get confused. When generating token $n+1$, if token n+1 is part of the $\frac{2n}{3}$, does it squeeze out an existing top-k token? Or do we dynamically expand the top-k as our sequence grows longer?

------
Also just to confirm, for expert routing during training and inference, for each filtered layer, do we only attend to the current layer, or do we also attend to previous depths? e.g. if token 1 stops at depth 1 but token 2 stops at depth 2, when computing attention for token 2 at depth 2, does it attend to token 1 at depth 1 as well, or it just pretends token 1 doesn't exist when computing token 2 depth 2?

Based on the following quote it sounds like we only stick to the current layer during training?
>  We observe that while recursive KV sharing offers the advantages of reduced memory footprint and overall FLOPs, it leads to quite large performance degradation in expert-choice routing under a fixed token setting. This suggest that exclusively updating and attending to the tokens active in that recursion depth may be more beneficial.

Thanks!
Max

**Comment by raymin0223** [comment]
Hi @maxyu1115 , thanks for your interests!

> This is where I get confused. When generating token  $n+1$, if token n+1 is part of the $\frac{2n}{3}$, does it squeeze out an existing top-k token? Or do we dynamically expand the top-k as our sequence grows longer?

Almost exactly! To understand inference behavior, I'd like to note Expert-choice routing has causality violation during training. This means to decide top-k tokens, we need to know all contexts in advance. In training, it works fine since the model gets all contexts at once. But in inference phase, the model will generate token by token, so we don't know the current token will be included in top-k.

This is why we and prior works for MoE use "auxiliary router" or "auxiliary loss". We can train additional extra router that learns if each token is included in top-k, through binary cross-entropy loss. Then we will remove the main MoR router during inference, and use this auxiliary router to decide recursion number for autoregressively generated tokens. 

Meanwhile, we can give a kind of regularization loss to original MoR router during training. Make it to output scores close to 1 for top-k tokens, and close to 0 for non-top-k tokens. If so, we can use this router even for inference. We give one single generated token to router, and apply recursion if the router outputs over 0.5.

> Also just to confirm, for expert routing during training and inference, for each filtered layer, do we only attend to the current layer, or do we also attend to previous depths? e.g. if token 1 stops at depth 1 but token 2 stops at depth 2, when computing attention for token 2 at depth 2, does it attend to token 1 at depth 1 as well, or it just pretends token 1 doesn't exist when computing token 2 depth 2?

We use "Recursion-wise KV caching" as our final design choice. In this setup, tokens only attend to activated tokens at each recursion step. So when computing attention for token 2 at depth 2, it does not attend to token 1 at depth 1. It will only see some activated tokens at depth 2.

---

We will try to make inference code for expert-choice when we have some spare times. And it would be good to see #9 and #10 issues as well!

**Comment by maxyu1115** [comment]
Thanks the explanations!
